File: app/src/main/java/in/woloo/www/woloo_host/mvp/CreateWolooHostPresenter.java
package in.woloo.www.woloo_host.mvp;

import android.app.Activity;
import android.content.Context;
import android.util.Log;
import in.woloo.www.utils.Logger;

import com.android.volley.VolleyError;
import com.google.gson.reflect.TypeToken;
import com.jetsynthesys.encryptor.JetEncryptor;

import org.json.JSONArray;
import org.json.JSONObject;

import java.lang.reflect.Type;
import java.util.List;

import in.woloo.www.common.CommonUtils;
import in.woloo.www.more.models.ViewProfileResponse;
import in.woloo.www.networksUtils.APIConstants;
import in.woloo.www.networksUtils.JSONTagConstant;
import in.woloo.www.networksUtils.NetworkAPICall;
import in.woloo.www.networksUtils.NetworkAPICallModel;
import in.woloo.www.networksUtils.NetworkAPIResponseCallback;
import in.woloo.www.utils.AppConstants;
import in.woloo.www.woloo_host.model.AddWolooResponse;
import in.woloo.www.woloo_host.model.GeoCodeResponse;

public class CreateWolooHostPresenter implements NetworkAPIResponseCallback {

    private static final String TAG = CreateWolooHostPresenter.class.getSimpleName();

    private Context mContext;
    private CreateWolooHostView createWolooHostView;
    private CommonUtils mCommonUtils;
    private JetEncryptor mJetEncryptor;
    private NetworkAPICall mNetworkAPICall;

    public CreateWolooHostPresenter(Context mContext, CreateWolooHostView createWolooHostView) {
        this.mContext = mContext;
        this.createWolooHostView = createWolooHostView;
        mCommonUtils = new CommonUtils();
        mJetEncryptor = JetEncryptor.getInstance();
        mNetworkAPICall = new NetworkAPICall();
    }

    public void getProfile(){
        try {
            JSONObject mJsObjParam = new JSONObject();
            Type parserType = new TypeToken<ViewProfileResponse>() {
            }.getType();
            NetworkAPICallModel networkAPICallModel = new NetworkAPICallModel(APIConstants.VIEW_PROFILE, AppConstants.POST_REQUEST, AppConstants.APP_TYPE_MOBILE, mJsObjParam, mJetEncryptor);
            networkAPICallModel.setParserType(parserType);
            networkAPICallModel.setShowProgress(true);
            mNetworkAPICall.callApplicationWS((Activity) mContext, networkAPICallModel, this);
        } catch (Exception e) {
            Logger.e(TAG, e.getMessage());
        }
    }

    public void getLocation(String lat,String lng){
        try {
            JSONObject mJsObjParam = new JSONObject();
            mJsObjParam.put("lat",lat);
            mJsObjParam.put("lng",lng);
            Type parserType = new TypeToken<GeoCodeResponse>() {
            }.getType();
            NetworkAPICallModel networkAPICallModel = new NetworkAPICallModel(APIConstants.GEO_CODE_LOCATION_API, AppConstants.POST_REQUEST, AppConstants.APP_TYPE_MOBILE, mJsObjParam, mJetEncryptor);
            networkAPICallModel.setParserType(parserType);
            networkAPICallModel.setShowProgress(true);
            mNetworkAPICall.callApplicationWS((Activity) mContext, networkAPICallModel, this);
        } catch (Exception e) {
            Logger.e(TAG, e.getMessage());
        }
    }

    public void becomeWolooHost(String name, String city, String address, String zipCode, JSONArray uploadedImages, double lat, double log){
        try {
            JSONObject mJsObjParam = new JSONObject();
            try{
                mJsObjParam.put(JSONTagConstant.NAME,name);
                mJsObjParam.put(JSONTagConstant.CITY,city);
                mJsObjParam.put(JSONTagConstant.ADDRESS,address + " " + city + " " + zipCode + " India");
                mJsObjParam.put(JSONTagConstant.IMAGE,uploadedImages);
                mJsObjParam.put(JSONTagConstant.ZIPCODE,zipCode);
                mJsObjParam.put(JSONTagConstant.LATITUDE,lat);
                mJsObjParam.put(JSONTagConstant.LONGITUDE,log);
            }catch (Exception ex){
                 CommonUtils.printStackTrace(ex);
            }
            Type parserType = new TypeToken<AddWolooResponse>() {
            }.getType();
            NetworkAPICallModel networkAPICallModel = new NetworkAPICallModel(APIConstants.ADD_WOLOO_HOST, AppConstants.POST_REQUEST, AppConstants.APP_TYPE_MOBILE, mJsObjParam, mJetEncryptor);
            networkAPICallModel.setParserType(parserType);
            networkAPICallModel.setShowProgress(true);
            mNetworkAPICall.callApplicationWS((Activity) mContext, networkAPICallModel, this);
        } catch (Exception e) {
            Logger.e(TAG, e.getMessage());
        }
    }

    @Override
    public void onSuccessResponse(JSONObject response, NetworkAPICallModel networkAPICallModel) {
        switch (networkAPICallModel.getApiURL()) {
            case APIConstants.VIEW_PROFILE:
                try {
                    ViewProfileResponse viewProfileResponse = (ViewProfileResponse) networkAPICallModel.getResponseObject();
                    if(viewProfileResponse != null){
                        viewProfileResponseFlow(viewProfileResponse);
                    }
                } catch (Exception e) {
                    mCommonUtils.printStackTrace(e);
                }
                break;
            case APIConstants.ADD_WOLOO_HOST:
                try{
                    AddWolooResponse addWolooResponse = (AddWolooResponse) networkAPICallModel.getResponseObject();
                    if(addWolooResponse != null){
                        addWolooHostResponseFlow(addWolooResponse);
                    }
                }catch (Exception ex){
                     CommonUtils.printStackTrace(ex);
                }
                break;
            case APIConstants.GEO_CODE_LOCATION_API:
                try{
                    GeoCodeResponse geoCodeResponse = (GeoCodeResponse) networkAPICallModel.getResponseObject();
                    if(geoCodeResponse != null){
                        geoCodeResponseFlow(geoCodeResponse);
                    }
                }catch (Exception ex){
                     CommonUtils.printStackTrace(ex);
                }
                break;
            default:
                break;
        }
    }

    private void geoCodeResponseFlow(GeoCodeResponse geoCodeResponse) {
        if(geoCodeResponse != null && geoCodeResponse.getStatus().equals(AppConstants.API_SUCCESS)){
            createWolooHostView.geoCodeResponseSuccess(geoCodeResponse);
        }
        else
            CommonUtils.showCustomDialogBackClick(mContext,"Unable to find Location Service. Please start your location Service Or Reboot your device.");

    }

    private void addWolooHostResponseFlow(AddWolooResponse addWolooResponse) {
        try{
            if(addWolooResponse != null && addWolooResponse.getData().getStatus().equals(AppConstants.API_SUCCESS)){
                createWolooHostView.addWolooHostSuccess(addWolooResponse.getData().getMessage());
            }
        }catch (Exception ex){
             CommonUtils.printStackTrace(ex);
        }
    }

    private void viewProfileResponseFlow(ViewProfileResponse viewProfileResponse) {
        try{
            if(viewProfileResponse != null && viewProfileResponse.getStatus().equals(AppConstants.API_SUCCESS)){
                createWolooHostView.setProfileResponse(viewProfileResponse);
            }
        }catch (Exception ex){
             CommonUtils.printStackTrace(ex);
        }
    }

    @Override
    public void onFailure(VolleyError volleyError, NetworkAPICallModel networkAPICallModel) {

    }

    @Override
    public void onNoInternetConnection(Activity context, NetworkAPICallModel networkAPICallModel, NetworkAPIResponseCallback networkAPIResponseCallback) {

    }

    @Override
    public void onTimeOutConnection(Activity context, NetworkAPICallModel networkAPICallModel, NetworkAPIResponseCallback networkAPIResponseCallback) {

    }
}

File: app/src/main/java/in/woloo/www/more/woloo_host/mvp/CreateWolooHostPresenter.kt
package `in`.woloo.www.more.woloo_host.mvp

import android.app.Activity
import android.content.Context
import com.android.volley.VolleyError
import com.google.gson.reflect.TypeToken
import com.jetsynthesys.encryptor.JetEncryptor
import `in`.woloo.www.application_kotlin.api_classes.APIConstants
import `in`.woloo.www.application_kotlin.api_classes.JSONTagConstant
import `in`.woloo.www.application_kotlin.api_classes.NetworkAPICallModel
import `in`.woloo.www.application_kotlin.api_classes.NetworkAPIResponseCallback
import `in`.woloo.www.common.CommonUtils
import `in`.woloo.www.more.models.ViewProfileResponse
import `in`.woloo.www.more.woloo_host.model.AddWolooResponse
import `in`.woloo.www.more.woloo_host.model.GeoCodeResponse
import `in`.woloo.www.networksUtils.NetworkAPICall
import `in`.woloo.www.utils.AppConstants
import `in`.woloo.www.utils.Logger
import org.json.JSONArray
import org.json.JSONObject

class CreateWolooHostPresenter(
    private val mContext: Context,
    private val createWolooHostView: CreateWolooHostView
) : NetworkAPIResponseCallback {
    private val mCommonUtils: CommonUtils = CommonUtils()
    private val mJetEncryptor: JetEncryptor
    private val mNetworkAPICall: NetworkAPICall

    init {
        mJetEncryptor = JetEncryptor.getInstance()
        mNetworkAPICall = NetworkAPICall()
    }

    val profile: Unit
        get() {
            try {
                val mJsObjParam = JSONObject()
                val parserType = object : TypeToken<ViewProfileResponse?>() {}.type
                val networkAPICallModel = NetworkAPICallModel(
                    APIConstants.VIEW_PROFILE,
                    AppConstants.POST_REQUEST,
                    AppConstants.APP_TYPE_MOBILE,
                    mJsObjParam,
                    mJetEncryptor
                )
                networkAPICallModel.parserType = parserType
                networkAPICallModel.isShowProgress = true
                mNetworkAPICall.callApplicationWS(mContext as Activity, networkAPICallModel, this)
            } catch (e: Exception) {
                Logger.e(TAG, e.message.toString())
            }
        }

    fun getLocation(lat: String?, lng: String?) {
        try {
            val mJsObjParam = JSONObject()
            mJsObjParam.put("lat", lat)
            mJsObjParam.put("lng", lng)
            val parserType = object : TypeToken<GeoCodeResponse?>() {}.type
            val networkAPICallModel = NetworkAPICallModel(
                APIConstants.GEO_CODE_LOCATION_API,
                AppConstants.POST_REQUEST,
                AppConstants.APP_TYPE_MOBILE,
                mJsObjParam,
                mJetEncryptor
            )
            networkAPICallModel.parserType = parserType
            networkAPICallModel.isShowProgress = true
            mNetworkAPICall.callApplicationWS(mContext as Activity, networkAPICallModel, this)
        } catch (e: Exception) {
            Logger.e(TAG, e.message.toString())
        }
    }

    fun becomeWolooHost(
        name: String?,
        city: String,
        address: String,
        zipCode: String,
        uploadedImages: JSONArray?,
        lat: Double,
        log: Double
    ) {
        try {
            val mJsObjParam = JSONObject()
            try {
                mJsObjParam.put(JSONTagConstant.NAME, name)
                mJsObjParam.put(JSONTagConstant.CITY, city)
                mJsObjParam.put(JSONTagConstant.ADDRESS, "$address $city $zipCode India")
                mJsObjParam.put(JSONTagConstant.IMAGE, uploadedImages)
                mJsObjParam.put(JSONTagConstant.ZIPCODE, zipCode)
                mJsObjParam.put(JSONTagConstant.LATITUDE, lat)
                mJsObjParam.put(JSONTagConstant.LONGITUDE, log)
            } catch (ex: Exception) {
                CommonUtils.printStackTrace(ex)
            }
            val parserType = object : TypeToken<AddWolooResponse?>() {}.type
            val networkAPICallModel = NetworkAPICallModel(
                APIConstants.ADD_WOLOO_HOST,
                AppConstants.POST_REQUEST,
                AppConstants.APP_TYPE_MOBILE,
                mJsObjParam,
                mJetEncryptor
            )
            networkAPICallModel.parserType = parserType
            networkAPICallModel.isShowProgress = true
            mNetworkAPICall.callApplicationWS(mContext as Activity, networkAPICallModel, this)
        } catch (e: Exception) {
            Logger.e(TAG, e.message.toString())
        }
    }

    override fun onSuccessResponse(
        response: JSONObject?,
        networkAPICallModel: NetworkAPICallModel?
    ) {
        when (networkAPICallModel!!.apiURL) {
            APIConstants.VIEW_PROFILE -> try {
                val viewProfileResponse: ViewProfileResponse =
                    networkAPICallModel.responseObject as ViewProfileResponse
                if (viewProfileResponse != null) {
                    viewProfileResponseFlow(viewProfileResponse)
                }
            } catch (e: Exception) {
                CommonUtils.printStackTrace(e)
            }

            APIConstants.ADD_WOLOO_HOST -> try {
                val addWolooResponse: AddWolooResponse =
                    networkAPICallModel.responseObject as AddWolooResponse
                if (addWolooResponse != null) {
                    addWolooHostResponseFlow(addWolooResponse)
                }
            } catch (ex: Exception) {
                CommonUtils.printStackTrace(ex)
            }

            APIConstants.GEO_CODE_LOCATION_API -> try {
                val geoCodeResponse = networkAPICallModel.responseObject as GeoCodeResponse
                geoCodeResponse?.let { geoCodeResponseFlow(it) }
            } catch (ex: Exception) {
                CommonUtils.printStackTrace(ex)
            }

            else -> {}
        }
    }

    private fun geoCodeResponseFlow(geoCodeResponse: GeoCodeResponse?) {
        if (geoCodeResponse != null && geoCodeResponse.status == AppConstants.API_SUCCESS) {
            createWolooHostView.geoCodeResponseSuccess(geoCodeResponse)
        } else CommonUtils.showCustomDialogBackClick(
            mContext,
            "Unable to find Location Service. Please start your location Service Or Reboot your device."
        )
    }

    private fun addWolooHostResponseFlow(addWolooResponse: AddWolooResponse?) {
        try {
            if (addWolooResponse != null && addWolooResponse.data!!.status == AppConstants.API_SUCCESS) {
                createWolooHostView.addWolooHostSuccess(addWolooResponse.data!!.message)
            }
        } catch (ex: Exception) {
            CommonUtils.printStackTrace(ex)
        }
    }

    private fun viewProfileResponseFlow(viewProfileResponse: ViewProfileResponse?) {
        try {
            if (viewProfileResponse != null && viewProfileResponse.status == AppConstants.API_SUCCESS) {
                createWolooHostView.setProfileResponse(viewProfileResponse)
            }
        } catch (ex: Exception) {
            CommonUtils.printStackTrace(ex)
        }
    }

    override fun onFailure(volleyError: VolleyError?, networkAPICallModel: NetworkAPICallModel?) {}
    override fun onNoInternetConnection(
        context: Activity?,
        networkAPICallModel: NetworkAPICallModel?,
        networkAPIResponseCallback: NetworkAPIResponseCallback?
    ) {
    }

    override fun onTimeOutConnection(
        context: Activity?,
        networkAPICallModel: NetworkAPICallModel?,
        networkAPIResponseCallback: NetworkAPIResponseCallback?
    ) {
    }

    companion object {
        private val TAG = CreateWolooHostPresenter::class.java.simpleName
    }
}
